lib/puppet/settings.rb

```diff
@@ -146,8 +147,8 @@ def initialize
     @configuration_file = nil
 
     # And keep a per-environment cache
-    @cache = Hash.new { |hash, key| hash[key] = {} }
-    @values = Hash.new { |hash, key| hash[key] = {} }
+    @cache = Concurrent::Hash.new { |hash, key| hash[key] = Concurrent::Hash.new }
+    @values = Concurrent::Hash.new { |hash, key| hash[key] = Concurrent::Hash.new }
```

Comment:
I think this may not work as expected when environment timeout is enabled, because if a thread marks an environment as expired, then only that thread's cache will be cleared. Requests handled by a different thread will still see the old realpath-ed `modulepath`, I think?

Reply:
I think this is just a [synchronized](https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/hash.rb#L8-L11) hash, not a separate hash per thread. But maybe I'm misunderstanding.

Reply:
Yes, this is just preventing simultaneous writes to the same hash from causing corruption. It doesn't change anything about the way this is used. I was getting errors when two threads resolved some interpolated value at the same time and they both tried to write to the cache.

Reply:
Ah, right makes sense. I saw `Concurrent` and my brain associated that with the thread local context.